doc/developer/design/20200520_exactly_once_kafka.md

```diff
@@ -1,28 +1,79 @@
+6. On startup, sink should read the latest complete timestamp from the consistency topic and drop any writes coming through the dataflow until caught up.
+
+### Mz changes (but with even more detail!)
+
+1. Minting timestamps consistently across all instances of a RT source is likely the hardest part of this project. Currently, RT sources:
+   a. Don't communicate any `upper` or `since` information back to the Coordinator.
+   b. Handle their timestamping logic in source operator local state. Each source operator, when invoked, checks the last time it updated the RT timestamp, and if that update happened before `now() - timestamp_frequency` it updates the RT timestamp. All records it receives from now until the end of the current invocation get assigned that timestamp. Rinse and repeat.
+
+This protocol is simple, and requires that each source instance hold a constant amount of data (the current RT timestamp) in memory. Unfortunately, it also means that queries like:
+
+```sql
+
+# Create RT Kafka source
+CREATE SOURCE foo from KAFKA BROKER ...
+
+# Create two source instances
+CREATE MATERIALIZED VIEW test1 AS SELECT * FROM foo;
+CREATE MATERIALIZED VIEW test2 AS SELECT * FROM foo;
+
+SELECT * FROM test1 EXCEPT ALL SELECT * FROM test2;
+```
+
+can return data, when logically they shouldn't.
+
+We will change this behavior and instead force all instances consuming from a given `(source_id, partition_id)` to read from the same worker, and then keep state about "what data have beeen assigned to what timestamps" in worker-local state, rather than operator local state.
+
+The worker will track, for each partition, a list of timestamp bindings, that are conceptually a mapping from `timestamp` to intervals `[start_offset, end_offset)`. When a source operator needs to timestamp a record for which there is no existing binding, it will propose a new binding from `current_timestamp` -> `[start_offset, end_offset')`. Periodically, the dataflow worker will finalize proposed bindings, and update the current timestamp and introduce a new timestamp binding for `new_timestamp -> [greatest_read_offset, greatest_read_offset)` that source operators can then propose new additions to.
+
+Since all the source operators are also invoked by the same worker, only one thing can happen at a time and there are no race conditions to worry about.
+
+Concretely, every time a source operator is invoked, it needs to:
+ - Check the worker's timestamp bindings for the next potential offset that data might appear at, so that the operator can respect the timestamp assignments of its peers
+ - Downgrade its capability based on the lowest available timestamp across all of the partitions that operator has been assigned to. If the operator has not been assigned to any timestamps it still needs to downgrade its capability to the `current_timestamp`. We cannot close the source operator in this case because the worker may be assigned to a new partition in the future.
```

Comment:
"If the operator has not been assigned to any timestamps" -> "If the operator has not been assigned to any parts/partitions"?